### ISSUE TITLE
Key-mode, structured key-mode

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/pumpp#71]
Similar to the `chord` transformers, there should be task transformers for key/mode estimation.

For consistency, we should also provide a structured key encoding that translates key and mode names into `(root, pitch classes)`.

**Comment by tomxi** [comment]
I'll officially claim this one now :)

the `KeyTransformer` will consume `key_mode` jams annotations and spit out `(pitch_profile, tonic)`

**Comment by tomxi** [comment]
Some design choices: (well really only 1)
When the `KeyTransformer` gets an 'N' label, it produces a `pitch_profile` that's all zeros as opposed to all ones. All ones would make sense if thinking of these things as masks, as in all pitches are fair game in atonal sections. But I did all zeros so now the masks mean only these pitches are more salient in tonal sections.

**Comment by bmcfee** [comment]
>     * When the `KeyTransformer` gets an 'N' label, the current version produces a `pitch_profile` that's all zeros as opposed to all ones.
>     * All ones would make sense if thinking of these things as masks, as in all pitches are fair game in atonal sections. But I did all zeros so now the masks mean only these pitches are more salient in tonal sections.

Good thinking.  Here are my thoughts:

Masks are one way to think about this representation, but another way is as exposing similarity between different keys.  This gets right to the assumptions of our model / annotations: what does `N` actually mean in our data?  What about `X`?  My guess is that annotators aren't too careful about their use of `N`, and that it could represent general atonality, non-functional sections, silence/noise, etc.

Since we don't need a strict interpretation of `N` in terms of this 12-bit vector, I'm okay with setting it to all-zeros and calling it a day.  It might be helpful to distinguish it from `X` by using all-ones, with the interpretation being that an out-of-gamut key can have any set of pitch classes active, but this will be a pretty loose interpretation.

>     * Another issue with all 0's is that the vector won't normalize in a probability sense... If later we want to turn these mask-like things to probabilities, all 1's makes more sense. But I ended up with all zeros.

The pitch profile outputs don't need to form a distribution / normalize properly.  If we think of each pitch class as independently being active or not, then it could make sense to have fractional membership (re your next observation), but all zeros or all ones should be totally fine.



>     * A related issue is the `dtype` of these `pitch_profile` vectors. Currently, they are `np.bool`s. This does not allow for non-mask information type like probability, this prevents the implementation of variations stemming from the discussion we had together with @lostanlen.

I see no problem in switching it out to floats, but you might hit some snags in training with keras.  IIRC rachel and I had to roll our own fractional cross-entropy loss to make that work in deep salience.

OTOH, since these don't necessarily have to normalize, it seems valid to me to have both m7 and M7 set to 1 for minor keys.  These are just structuring hints for the model, after all, and not hard-and-fast rules.



>     * For minor mode, the profile reflects the harmonic minor. In other words, C Aeolian and C minor have the same profile.
>     * If no mode string follows the main tonic, the default profile is for major or Ionian.


That's what I would expect. :+1:

It does raise some ambiguity in the decoder, but I'm not too worried about it.  I very much doubt we have any annotations that say Ionian instead of major.

**Comment by bmcfee** [comment]
Fixed by merging #111
